### PR TITLE
JENKINS-35489 - Remove Bouncy Castle dependency from instance identity module

### PR DESCRIPTION
**Pull request by alvarolobato**
[JENKINS-35489](https://issues.jenkins-ci.org/browse/JENKINS-35489)
Dependency to multiple Bouncy Castle versions from jenkins core and plugins is causing problems due to the binary incompatibility between versions, the different supported algorithms, etc.

Plugins should be migrated to use a common API and BC version in order to avoid these problems and the dependency from BC should be removed from the core.

This PR removes that dependency from the core by migrating the instance identity key from PCKS1 encoding to PCKS8 which is natively supported by the JDK.

After this PR, instance identity module will be able to read any instance identity key encoded in PCKS1 or PCKS8 and will always write it in PCKS8 format.

see [bouncycastle-api plugin](https://github.com/jenkinsci/bouncycastle-api/blob/master/README.md)

@reviewbybees esp: @stephenc

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -13,6 +13,10 @@
   <name>Instance Identity</name>
   <description>Maintains an RSA key pair that can serve as a foundation of authentication when communicating with Jenkins</description>
 
+  <properties>
+    <jenkins.version>2.7</jenkins.version>
```
Comment (amuniz) [inline pom.xml]: Why are you bumping core from `1.609.1`?

File: pom.xml [hunk 2]
```diff
@@ -28,9 +32,10 @@
 
   <dependencies>
     <dependency>
-      <groupId>org.bouncycastle</groupId>
-      <artifactId>bcpkix-jdk15on</artifactId>
-      <version>1.54</version>
+       <groupId>org.jenkins-ci.plugins</groupId>
+       <artifactId>bouncycastle-api</artifactId>
```
Comment (amuniz) [inline pom.xml]: Maybe I'm missing something, but I don't see any use of this API in this PR.

Comment (alvarolobato) [inline pom.xml]: Only in the tests to make sure that what we read and write is consistent with what BC would do:
https://github.com/jenkinsci/instance-identity-module/pull/6/files#diff-12d629d9d550350a15b7ba219656ba62R140

File: pom.xml (hunk 1 above)
Comment (alvarolobato) [inline pom.xml]: It was a test I forgot to delete. Corrected.

File: pom.xml (hunk 2 above)
Comment (oleg-nenashev) [inline pom.xml]: Seems it also bumps the JDK dependency. Most likely it's fine

**Comment by oleg-nenashev** [comment]
🐛 Jenkins modules should not depend on plugins. By doing it you effectively bundle the plugin codebase inside the core. In such case any plugin updates are not going to work and may effectively blow up the instance

**Comment by alvarolobato** [comment]
@oleg-nenashev as we have just talked, I'm going to change the test relying on static files and avoid having to use bouncycastle or the api plugin to check the results.

**Comment by alvarolobato** [comment]
@oleg-nenashev I've removed the dependency and reworked the tests.

**Comment by stephenc** [comment]
:bug: bump version in main pom to 2.0 as you have dropped a dependency (yes we want to drop the dependency, but we should bump the major for this change to highlight that bouncycastle is not longer transitive from instance-identity)

**Comment by stephenc** [comment]
otherwise "All good"

**Comment by alvarolobato** [comment]
@stephenc version bump done.

**Comment by oleg-nenashev** [comment]
:ant: Needs some polishing of LICENSE.txt and pom.xml, because now the JAR also includes much apache code.

Would be better if the code just depends on the shaded library version, but oh well.
We also should ensure that the 2.0 bump does not break anything in Jenkins core and plugins before merging. I suspect it does :(

**Comment by stephenc** [comment]
The LICENSE of this code is fine as MIT, but we now need a NOTICE.txt that gets included in the .jar due to including Apache Licensed code.

@oleg-nenashev the core version bump will be aligned with an implied dependency on bouncycastle-api

**Comment by oleg-nenashev** [comment]
@stephenc But how are you going to handle plugins, which transitively depend on the core. We have no solution in plugin loader, which would reroute such obsolete dependencies to plugin references and then perform dynamic load

**Comment by stephenc** [comment]
@oleg-nenashev we will have to make a fix **when** we merge into core the newer instance-identity module. For now the first change will be to merge instance-identity and promote the bouncy-castle dependency... then when we have a solution we will shift the dependency into one on the bouncy-castle-api just as for any standard decoupling of core classes from core... same solution as for all of those decouplings

**Comment by oleg-nenashev** [comment]
🐝 with some :aware: and NIT regarding code bundling

**Comment by alvarolobato** [comment]
@stephenc is this a bee now?

**Comment by stephenc** [comment]
:bee:

**Comment by alvarolobato** [comment]
@reviewbybees done

**Comment by alvarolobato** [comment]
@oleg-nenashev or @stephenc can you merge this PR?

**Pull request merged**